datastore/datastore/storagebackend/postgresql/putobservations.go

```diff
@@ -63,32 +70,68 @@ func getTSColVals(tsMdata *datastore.TSMetadata) ([]interface{}, error) {
 		if method.IsValid() {
 			val, ok := method.Call([]reflect.Value{})[0].Interface().(string)
 			if !ok {
-				return nil, fmt.Errorf(
+				return nil, nil, fmt.Errorf(
 					"method.Call() failed for method %s; failed to return string", methodName)
 			}
 			colVals = append(colVals, val)
+			colName2Val[common.ToSnakeCase(field.Name)] = val
 		}
 	}
 
 	// --- END string metadata ---------------------------
 
-	return colVals, nil
+	// derive colValsUnique from colName2Val
+	colValsUnique, err := getTSColValsUnique(colName2Val)
+	if err != nil {
+		return nil, nil, fmt.Errorf("getTSColValsUnique() failed: %v", err)
+	}
+
+	return colVals, colValsUnique, nil
 }
 
-// getTimeSeriesID retrieves the ID of the row in table time_series that matches tsMdata,
-// inserting a new row if necessary. The ID is first looked up in a cache in order to save
-// unnecessary database access.
+// getTSColValsUnique gets the subset of colName2Val that correspond to the fields defined by
+// constraint unique_main.
+//
+// The order in the returned array is consistent with the array returned by getTSColNamesUnique().
+//
+// Returns (column values, nil) upon success, otherwise (..., error).
+func getTSColValsUnique(colName2Val map[string]interface{}) ([]interface{}, error) {
+
+	result := []interface{}{}
+
+	for _, col := range getTSColNamesUnique() {
+		colVal, found := colName2Val[col]
+		if !found {
+			return []interface{}{},
+				fmt.Errorf("column '%s' not found in colName2Val: %v", col, colName2Val)
+		}
+		result = append(result, colVal)
+	}
+
+	return result, nil
+}
+
+// upsertTS retrieves the ID of the row in table time_series that matches tsMdata wrt.
+// the fields - U - defined by constraint unique_main, inserting a new row if necessary.
+//
+// If the row already existed, the function ensures that the row is updated with the tsMdata
+// fields - UC - that are not in U (i.e. the complement of U).
+//
+// The ID is first looked up in a cache (where the key consists of all fields (U + UC)) in order to
+// save unnecessary database access. In other words, a cache hit means that the row for
+// time series not only existed, but was also already fully updated according to tsMdata.
+// And vice versa: a cache hit means the row either didn't exist at all or wasn't fully updated
```

Comment:
miss?

Reply:
Definitely!

Reply:
Will piggyback this fix on the next PR. 🐖